spec/abstract/Sporadic_A.thy

```diff
@@ -79,4 +80,42 @@ where
     od
   od"
 
+(* below currently gives if-test (when) for assert *)
+definition if_cond_refill_unblock_check where
+  "if_cond_refill_unblock_check sc_opt active asst \<equiv>
+    maybeM (\<lambda>scp. do
+      sc \<leftarrow> get_sched_context scp;
+      cur_sc_ptr \<leftarrow> gets cur_sc;
+      guard \<leftarrow> return (case active of
+                         None \<Rightarrow> (\<not> sc_sporadic sc)
+                       | Some True \<Rightarrow> sc_sporadic sc \<and> sc_active sc
+                       | Some False \<Rightarrow> sc_sporadic sc \<and> sc_active sc);
+      when (guard \<and> (if asst = Some False then scp \<noteq> cur_sc_ptr else True)) $
+        when (if asst = Some True then scp \<noteq> cur_sc_ptr else True) $ refill_unblock_check scp
```

Comment:
Does this mirror a specific code structure in C/Haskell? Usually I'd prefer to write `when (A /\ B)` to `when A $ when B`.

Reply:
Yes, this is meant to be mirroring the C code. When `asst` is `Some False`, it always shows up as a conjunct in the first guard. But when `asst` is `Some True`, the “if without else (= when)” always sits inside the first guard, and originally this was an assert rather than a test - we changed it to a test to defer proving the asserts, making this case do the same test as the `Some False` case. But the structure of the C code remains the same, so I am keeping it here and in Haskell just to save the effort of changing now and also in refinement.

There is also some repetition with the `act` parameter, too. This is again meant to be two different guard, `sporadic` and `sporadic & active` tests, but since we are deferring of the proof of “sporadic implies active” for those call sites, the `sporadic` test is currently replaced with `sporadic & active`, which makes the two the same, but the original structure is kept (meaning the duplication is still there) because in C these correspond to slightly different codes and also because it should be changed back in the future.

`refill_unblock_check` is typically called under certain conditions, and there are a few patterns for those guards. This function aims to package those conditions up in an effective way so that we can make necessary arguments without unfolding it. I have written (and might add a few more) some wp rules that should cater for what we need for the cases where the difference in the parameters matters.

So, these were my thinking for keeping the original C structure as much as possible in these definitions. But that is based on the assumption that this will become useful when we revert these changes for “ease of verification” in the “near” future. If that’s unlikely, maybe we can just embrace these changes and optimise things in that way, including the C code. But we might want to do that separately, because we’d want to be merging this PR soon so that we can continue with the `schedule_corres` work.